src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_form.ts

```diff
@@ -93,34 +107,52 @@ export function useForm<T extends object = FormData>(
           {} as T
         );
 
-  const updateFormDataAt: FormHook<T>['__updateFormDataAt'] = (path, value) => {
-    const currentFormData = formData$.current.value;
-    formData$.current.next({ ...currentFormData, [path]: value });
-    return formData$.current.value;
+  const isFieldValid = (field: FieldHook) =>
+    field.getErrorsMessages() === null && !field.isValidating;
+
+  const updateFormValidity = () => {
+    const fieldsArray = fieldsToArray();
+    const areAllFieldsValidated = fieldsArray.every(field => field.isValidated);
+
+    if (!areAllFieldsValidated) {
+      // If *not* all the fiels have been validated, the validity of the form is unknown, thus still "undefined"
+      return;
+    }
+
+    const isFormValid = fieldsArray.every(isFieldValid);
+
+    setIsValid(isFormValid);
+    return isFormValid;
   };
 
-  /**
-   * When a field value changes, validateFields() is called with the field name + any other fields
-   * declared in the "fieldsToValidateOnChange" (see the field config).
-   *
-   * When this method is called _without_ providing any fieldNames, we only need to validate fields that are pristine
-   * as the fields that are dirty have already been validated when their value changed.
-   */
   const validateFields: FormHook<T>['__validateFields'] = async fieldNames => {
     const fieldsToValidate = fieldNames
-      ? fieldNames.map(name => fieldsRefs.current[name]).filter(field => field !== undefined)
-      : fieldsToArray().filter(field => field.isPristine); // only validate fields that haven't been changed
+      .map(name => fieldsRefs.current[name])
+      .filter(field => field !== undefined);
 
-    const formData = getFormData({ unflatten: false });
+    if (fieldsToValidate.length === 0) {
+      // Nothing to validate
+      return true;
+    }
 
+    const formData = getFormData({ unflatten: false });
     await Promise.all(fieldsToValidate.map(field => field.validate({ formData })));
+    updateFormValidity();
 
-    const isFormValid = fieldsToArray().every(
-      field => field.getErrorsMessages() === null && !field.isValidating
-    );
-    setIsValid(isFormValid);
+    return fieldsToValidate.every(isFieldValid);
+  };
 
-    return isFormValid;
+  const validateAllFields = async (): Promise<boolean> => {
+    const fieldsToValidate = fieldsToArray().filter(field => !field.isValidated);
+
+    if (fieldsToValidate.length === 0) {
+      // Nothing left to validate, all fields are already validated.
+      return isValid!;
+    }
+
+    await validateFields(fieldsToValidate.map(field => field.path));
+
+    return updateFormValidity()!;
```

Comment:
> Can we just keep the one in validateFields and remove the one in validateAllFields?

Good catch, I will remove the one in `validateAllFields()`.

> This code looks suspiciously like it's checking that every field is valid. Yet updateFormValidity returns isFormValid. So in addition to making redundant function calls, are we also doing the same thing in two different ways?

`validateFields()` requires `fieldNames` to be provided and it returns the validity of _those_ fields. At the same time, it updates the _form_ validity in case it has changed.

> Ideally, it seems like there should be a single place where the form validity state is set

This is the case, the only place where the form validity is set is inside `updateFormValidity()`. It is also unset in the new `reset()` method.

> we might need to use a ref to avoid returning a stale value

Not sure why we would prefer a ref than a state. This is probably the most important state of the form 😊